File: backend/app/schemas/company.py
from datetime import datetime
from pydantic import BaseModel, EmailStr, Field
from app.models.company import BusinessActivity


class CompanyBase(BaseModel):
    # Section 1
    logo_url: str | None = None
    name: str = Field(min_length=1, max_length=200)
    legal_structure_code: str | None = None
    primary_place_of_business: str | None = None
    business_activity: BusinessActivity | None = None

    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    district_code: str | None = None
    taluk_code: str | None = None
    pincode: str | None = None
    state: str = "Tamil Nadu"

    sector_code: str | None = None

    contact_name: str | None = None
    contact_designation: str | None = None
    contact_email: EmailStr | None = None
    contact_phone: str | None = None
    website: str | None = None

    workforce_count: int | None = Field(default=None, ge=0)
    turnover_range_code: str | None = None
    exact_turnover_lakhs: float | None = Field(default=None, ge=0)
    is_mnc: bool = False

    # Section 2
    gst_number: str | None = None
    cin: str | None = None
    udyam_number: str | None = None
    pan: str | None = None


class CompanyCreate(CompanyBase):
    tags: list[str] = []


class CompanyUpdate(BaseModel):
    logo_url: str | None = None
    name: str | None = Field(default=None, min_length=1, max_length=200)
    legal_structure_code: str | None = None
    primary_place_of_business: str | None = None
    business_activity: BusinessActivity | None = None

    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    district_code: str | None = None
    taluk_code: str | None = None
    pincode: str | None = None
    state: str | None = None

    sector_code: str | None = None

    contact_name: str | None = None
    contact_designation: str | None = None
    contact_email: EmailStr | None = None
    contact_phone: str | None = None
    website: str | None = None

    workforce_count: int | None = Field(default=None, ge=0)
    turnover_range_code: str | None = None
    exact_turnover_lakhs: float | None = Field(default=None, ge=0)
    is_mnc: bool | None = None

    gst_number: str | None = None
    cin: str | None = None
    udyam_number: str | None = None
    pan: str | None = None

    tags: list[str] | None = None


class CompanyOut(CompanyBase):
    id: int
    tags: list[str]
    owner_user_id: int | None
    created_at: datetime
    updated_at: datetime

    # Computed
    profile_completion: int  # 0-100
    section_completion: dict[str, bool]  # per-section flags

    class Config:
        from_attributes = True


class CompanyListItem(BaseModel):
    """Lightweight shape for the registry table."""
    id: int
    name: str
    gst_number: str | None
    sector_code: str | None
    district_code: str | None
    turnover_range_code: str | None
    tags: list[str]
    profile_completion: int
    is_mnc: bool


class CompanyListResponse(BaseModel):
    items: list[CompanyListItem]
    total: int
    limit: int
    offset: int


class LockedFieldsResponse(BaseModel):
    locked_for_msme: list[str]
    tag_edit_roles: list[str]

File: frontend/src/types/company.ts
export type BusinessActivity = 'manufacturing' | 'service' | 'trade'

export interface Company {
  id: number
  logo_url: string | null
  name: string
  legal_structure_code: string | null
  primary_place_of_business: string | null
  business_activity: BusinessActivity | null

  address_line1: string | null
  address_line2: string | null
  city: string | null
  district_code: string | null
  taluk_code: string | null
  pincode: string | null
  state: string

  sector_code: string | null

  contact_name: string | null
  contact_designation: string | null
  contact_email: string | null
  contact_phone: string | null
  website: string | null

  workforce_count: number | null
  turnover_range_code: string | null
  exact_turnover_lakhs: number | null
  is_mnc: boolean

  gst_number: string | null
  cin: string | null
  udyam_number: string | null
  pan: string | null

  tags: string[]
  owner_user_id: number | null
  created_at: string
  updated_at: string

  profile_completion: number
  section_completion: Record<string, boolean>
}

export interface CompanyListItem {
  id: number
  name: string
  gst_number: string | null
  sector_code: string | null
  district_code: string | null
  taluk_code: string | null
  pincode: string | null
  turnover_range_code: string | null
  tags: string[]
  profile_completion: number
  is_mnc: boolean
}

export interface CompanyListResponse {
  items: CompanyListItem[]
  total: number
  limit: number
  offset: number
}

export interface CompanyCreate {
  logo_url?: string | null
  name: string
  legal_structure_code?: string | null
  primary_place_of_business?: string | null
  business_activity?: BusinessActivity | null
  address_line1?: string | null
  address_line2?: string | null
  city?: string | null
  district_code?: string | null
  taluk_code?: string | null
  pincode?: string | null
  state?: string
  sector_code?: string | null
  contact_name?: string | null
  contact_designation?: string | null
  contact_email?: string | null
  contact_phone?: string | null
  website?: string | null
  workforce_count?: number | null
  turnover_range_code?: string | null
  exact_turnover_lakhs?: number | null
  is_mnc?: boolean
  gst_number?: string | null
  cin?: string | null
  udyam_number?: string | null
  pan?: string | null
  tags?: string[]
}

export type CompanyUpdate = Partial<CompanyCreate>

export interface CompanyFilters {
  q?: string
  sector?: string
  district?: string
  pincode?: string
  turnover?: string
  tag?: string
  limit?: number
  offset?: number
}

export interface LockedFields {
  locked_for_msme: string[]
  tag_edit_roles: string[]
}
